core/src/main/java/com/linecorp/armeria/server/RouteBuilder.java

```diff
@@ -0,0 +1,351 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.linecorp.armeria.internal.ArmeriaHttpUtil.concatPaths;
+import static com.linecorp.armeria.internal.RouteUtil.EXACT;
+import static com.linecorp.armeria.internal.RouteUtil.GLOB;
+import static com.linecorp.armeria.internal.RouteUtil.PREFIX;
+import static com.linecorp.armeria.internal.RouteUtil.REGEX;
+import static com.linecorp.armeria.internal.RouteUtil.ensureAbsolutePath;
+import static com.linecorp.armeria.server.HttpHeaderUtil.ensureUniqueMediaTypes;
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import javax.annotation.Nullable;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.MediaType;
+
+/**
+ * Builds a new {@link Route}.
+ */
+public class RouteBuilder {
+
+    @Nullable
+    private PathMapping pathMapping;
+
+    private Set<HttpMethod> methods = ImmutableSet.of();
+
+    private List<MediaType> consumes = ImmutableList.of();
+
+    private List<MediaType> produces = ImmutableList.of();
+
+    /**
+     * Sets the specified {@code pathPattern} of the {@link Route}. e.g.
+     * <ul>
+     *   <li>{@code /login} (no path parameters)</li>
+     *   <li>{@code /users/{userId}} (curly-brace style)</li>
+     *   <li>{@code /list/:productType/by/:ordering} (colon style)</li>
+     *   <li>{@code exact:/foo/bar} (exact match)</li>
+     *   <li>{@code prefix:/files} (prefix match)</li>
+     *   <li><code>glob:/~&#42;/downloads/**</code> (glob pattern)</li>
+     *   <li>{@code regex:^/files/(?<filePath>.*)$} (regular expression)</li>
+     * </ul>
+     *
+     * @throws IllegalArgumentException if the specified path pattern is invalid
+     */
+    public RouteBuilder path(String pathPattern) {
+        return pathMapping(getPathMapping(pathPattern));
+    }
+
+    private static PathMapping getPathMapping(String pathPattern) {
```

Comment:
Class would probably be more readable if all the public methods are on top.

Reply:
I prefer these sort of private methods are located right under the method where they are used. But at the bottom is fine too. Let me move it down.

Reply:
Yeah it does make the code implementation easier to read - but I think a lot of people read the class as documentation (in lieu of reading the javadoc) so having all the publics together helps for that :)

Reply:
Ah, yeah didn't think of that. :-) It could be an obstacle especially when the private method is too long.